Add `loki.source.api` component (#3648)
grafana/agent: docs/sources/flow/reference/components/loki.source.api.md

`loki.source.api` listens for HTTP requests containing Loki log entries and forwards them to other components capable
of receiving log entries.

The HTTP API exposed is compatible with [Loki push API][loki-push-api], using Loki's `logproto` format. This means that
other [`loki.write`][loki.write] components can be used as a client and send requests to `loki.source.api`. This enables
Grafana Agents to send log entries over the network to other Grafana Agents.

[loki.write]: {{< relref "./loki.write.md" >}}
[loki-push-api]: https://grafana.com/docs/loki/latest/api/#push-log-entries-to-loki

## Usage

```river
loki.source.api LABEL {
    http_address = "LISTEN_ADDRESS"
    http_port = PORT
    forward_to = RECEIVER_LIST
}
```

## Arguments

`loki.source.api` supports the following arguments:

 Name                     | Type                 | Description                                                | Default | Required 
--------------------------|----------------------|------------------------------------------------------------|---------|----------
 `http_port`              | `int`                | The host port for the HTTP server to listen on.            | `0`     | yes      

>  `http_address`           | `string`             | The host address for the HTTP server to listen on.         | `""`    | yes      

Reviewer: Since these fields are required, let's leave the Default empty.

 `forward_to`             | `list(LogsReceiver)` | List of receivers to send log entries to.                  |         | yes      
 `use_incoming_timestamp` | `bool`               | Whether or not to use the timestamp received from request. | `false` | no       
 `labels`                 | `map(string)`        | The labels to associate with each received logs record.    | `{}`    | no       
 `relabel_rules`          | `RelabelRules`       | Relabeling rules to apply on log entries.                  | `{}`    | no       

The `relabel_rules` field can make use of the `rules` export value from a
[`loki.relabel`][loki.relabel] component to apply one or more relabeling rules to log entries
before they're forwarded to the list of receivers in `forward_to`.

[loki.relabel]: {{< relref "./loki.relabel.md" >}}

## Exported fields

`loki.source.api` does not export any fields.

## Component health

`loki.source.api` is only reported as unhealthy if given an invalid configuration.

> ## Metrics

Author: There are more metrics that come with the weaveworks' server. I didn't think they are important enough to be on this list. Some other components don't mention them at all.

The metrics I've skipped:
```
# HELP loki_source_api_inflight_requests Current number of inflight requests.
# HELP loki_source_api_tcp_connections_limit The max number of TCP connections that can be accepted (0 means no limit).
```

Happy to try unify these, but again, this could be something for the Server unification.

The following are some of the metrics that are exposed when this component is used. Note that the metrics include labels
such as `status_code` where relevant, which can be used to measure request success rates.

* `loki_source_api_request_duration_seconds` - Time (in seconds) spent serving HTTP requests (histogram).
* `loki_source_api_request_message_bytes` - Size (in bytes) of messages received in the request (histogram).
* `loki_source_api_response_message_bytes` - Size (in bytes) of messages sent in response (histogram).
* `loki_source_api_tcp_connections` - Current number of accepted TCP connections (gauge).

## Example

This example starts an HTTP server on `localhost` address and port `9999`. The server receives log entries and forwards
them to a `loki.echo` component while
adding a `forwarded="true"` label.

```river
loki.echo "print" {}

loki.source.api "loki_push_api" {

>     http_address = "localhost"

Reviewer: Since people will be copying this example, should we make the listen address here `0.0.0.0` so the component can also receive requests from other machines over the network? (since I guess this will be its primary use case)

Author: Good idea!

---
title: loki.source.api
---

> # loki.source.api

Reviewer: Let's find some place to mention what the two exposed endpoints are and what different purpose they serve as currently people wouldn't know where to send logs to.

Reviewer: There's a very concise definition [here](https://grafana.com/docs/loki/next/clients/promtail/configuration/#loki_push_api)

```
Promtail also exposes a second endpoint on /promtail/api/v1/raw which expects newline-delimited log lines. This can be used to send NDJSON or plaintext logs.
```

Also, from the codebase I think we could say that the raw endpoint does not honor `use_incoming_timestamp`.

Author: That will do. Thanks!

> `loki.source.api` listens for HTTP requests containing Loki log entries and forwards them to other components capable of receiving log entries.

Reviewer: How does that sound?

```suggestion
`loki.source.api` receives log entries over HTTP and forwards them to other `loki.*` components.
```

> The HTTP API exposed is compatible with [Loki push API][loki-push-api], using Loki's `logproto` format. This means that other [`loki.write`][loki.write] components can be used as a client and send requests to `loki.source.api`. This enables Grafana Agents to send log entries over the network to other Grafana Agents.

Reviewer: ```suggestion
The HTTP API exposed is compatible with [Loki push API][loki-push-api] and the `logproto` format. This means that other [`loki.write`][loki.write] components can be used as a client and send requests to `loki.source.api` and enables using the Agent as a proxy for logs.
```

[loki.write]: {{< relref "./loki.write.md" >}}
[loki-push-api]: https://grafana.com/docs/loki/latest/api/#push-log-entries-to-loki

## Usage

```river

> loki.source.api LABEL {

Reviewer: ```suggestion
loki.source.api "LABEL" {
```

The `relabel_rules` field can make use of the `rules` export value from a
[`loki.relabel`][loki.relabel] component to apply one or more relabeling rules to log entries before they're forwarded to the list of receivers in `forward_to`.

[loki.relabel]: {{< relref "./loki.relabel.md" >}}

## Blocks

The following blocks are supported inside the definition of `loki.source.heroku`:

 Hierarchy | Name     | Description                                        | Required 
-----------|----------|----------------------------------------------------|----------
 `http`    | [http][] | Configures the HTTP server that receives requests. | no       

[http]: #http

### http

{{< docs/shared lookup="flow/reference/components/loki-server-http.md" source="agent" >}}

## Exported fields

`loki.source.api` does not export any fields.

## Component health

`loki.source.api` is only reported as unhealthy if given an invalid configuration.

## Debug metrics

The following are some of the metrics that are exposed when this component is used. Note that the metrics include labels such as `status_code` where relevant, which can be used to measure request success rates.

> * `agent_loki_source_api_request_duration_seconds` (histogram): Time (in seconds) spent serving HTTP requests.
> * `agent_loki_source_api_request_message_bytes` (histogram): Size (in bytes) of messages received in the request.
> * `agent_loki_source_api_response_message_bytes` (histogram): Size (in bytes) of messages sent in response.
> * `agent_loki_source_api_tcp_connections` (gauge): Current number of accepted TCP connections.

Reviewer: Let's remove the `agent_` prefix from our metrics (and fix that on our code as well) to conform with what other components are doing as well.

```suggestion
* `loki_source_api_request_duration_seconds` (histogram): Time (in seconds) spent serving HTTP requests.
* `loki_source_api_request_message_bytes` (histogram): Size (in bytes) of messages received in the request.
* `loki_source_api_response_message_bytes` (histogram): Size (in bytes) of messages sent in response.
* `loki_source_api_tcp_connections` (gauge): Current number of accepted TCP connections.
```

## Example

This example starts an HTTP server on `0.0.0.0` address and port `9999`. The server receives log entries and forwards them to a `loki.echo` component while adding a `forwarded="true"` label.

```river

> loki.echo "print" {}

Reviewer: nit: could we forward to a loki.write instead and amend the previous sentence? 

I think it will be more useful for people who are looking to copy/paste from our docs

```
loki.write "local" {
  endpoint {
    url = "loki:3100/api/v1/push"
    basic_auth {
      username = env("LOKI_USERNAME")
      password = env("LOKI_PASSWORD")
    }
  }
}
```
